contrib/IECoreUSD/src/IECoreUSD/PrimitiveAlgo.cpp

```diff
@@ -113,3 +114,122 @@ pxr::TfToken IECoreUSD::PrimitiveAlgo::toUSD( IECoreScene::PrimitiveVariable::In
 			return pxr::TfToken();
 	}
 }
+
+//////////////////////////////////////////////////////////////////////////
+// Reading primitive variables
+//////////////////////////////////////////////////////////////////////////
+
+namespace
+{
+
+void readPrimitiveVariable( const pxr::UsdGeomPrimvar &primVar, pxr::UsdTimeCode time, IECoreScene::Primitive *primitive )
+{
+	IECoreScene::PrimitiveVariable::Interpolation interpolation = IECoreUSD::PrimitiveAlgo::fromUSD( primVar.GetInterpolation() );
+	if( interpolation == IECoreScene::PrimitiveVariable::Invalid )
+	{
+		IECore::msg(IECore::MessageHandler::Level::Warning, "IECoreUSD::PrimitiveAlgo", boost::format( "Invalid Interpolation on %1%" ) % primVar.GetName().GetString() );
+		return;
+	}
+
+	pxr::VtValue value;
+	if( !primVar.Get( &value, time ) )
+	{
+		return;
+	}
+
+	IECore::DataPtr data = DataAlgo::fromUSD( value, primVar.GetTypeName() );
+	if( !data )
+	{
+		IECore::msg( IECore::MessageHandler::Level::Warning, "IECoreUSD::PrimitiveAlgo", boost::format( "PrimVar: %1% type: %2% not supported - skipping" ) % primVar.GetName().GetString() % primVar.GetTypeName() );
+		return;
+	}
+
+	pxr::VtIntArray srcIndices;
+	primVar.GetIndices( &srcIndices, time );
+	IECore::IntVectorDataPtr indices;
+	if( !srcIndices.empty() )
+	{
+		indices = DataAlgo::fromUSD( srcIndices );
+	}
+
+	primitive->variables[primVar.GetPrimvarName().GetString()] = IECoreScene::PrimitiveVariable( interpolation, data, indices );
+}
```

Comment:
No, `USDSceneTest.testPrimVarTypes` demonstrates that we can round-trip SimpleTypedData primitive variables through USD perfectly well. The "always an array even for constant interpolation" issue only applies to the special attributes like `width`. This is one of the aspects of USD that I've been less enamoured with. It would be much easier to deal with if things like the WidthsAttr API was a convenience on top of the generic primvars API, rather than a separate thing.